### PR TITLE
Use pureconfig for CouchDbRestStore

### PR DESCRIPTION
**Pull request by chetanmeh**
PR for #3355 Now the CouchDB config is read from application.conf with following structure

```
whisk {
    couchdb {
        protocol = "http"
        host     = "localhost"
        port     = 5984
        username = "user"
        password = "password"
        provider = "CouchDB"
        databases {
          WhiskAuth       = whisk_local_subjects
          WhiskEntity     = whisk_local_whisks
          WhiskActivation = whisk_local_activations
        }
    }
  }
```

The config can be provided via following environment variables

```
CONFIG_whisk_couchdb_protocol
CONFIG_whisk_couchdb_host
CONFIG_whisk_couchdb_port
CONFIG_whisk_couchdb_username
CONFIG_whisk_couchdb_password
CONFIG_whisk_couchdb_provider
CONFIG_whisk_couchdb_databases_WhiskAuth
CONFIG_whisk_couchdb_databases_WhiskEntity
CONFIG_whisk_couchdb_databases_WhiskActivation
```

This would be a backward incompatible change for existing deployments where they would need to switch to new environment variables instead of existing DB_PORT etc

~For now `ArtifactStoreProvider#makeStore` still takes WhiskConfig for determining DB name. That would be addressed in a separate PR~ now included in this PR itself

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 1]
```diff
@@ -36,21 +47,18 @@ object CouchDbStoreProvider extends ArtifactStoreProvider {
...
+    val dbConfig = loadConfigOrThrow[CouchDbConfig](ConfigKeys.couchdb)
     require(
-      config.dbProvider == "Cloudant" || config.dbProvider == "CouchDB",
-      "Unsupported db.provider: " + config.dbProvider)
-    assume(
-      Set(config.dbProtocol, config.dbHost, config.dbPort, config.dbUsername, config.dbPassword, name(config))
-        .forall(_.nonEmpty),
-      "At least one expected property is missing")
+      dbConfig.provider == "Cloudant" || dbConfig.provider == "CouchDB",
+      s"Unsupported db.provider: ${dbConfig.provider}")
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: Is there anything left that depends on `provider`? If not, please rip it out.

Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: Not sure what is the intention of provider so far. In code this param is not used. I just kept existing logic as is with switch to pureconfig. If provider is not required (if others can confirm) I can then remove that also

Comment (markusthoemmes) [reply common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: @rabbah can you confirm? I think we used to have a distinction but killed all of that long ago with a hard switch to just CouchDB APIs.

**Comment by chetanmeh** [comment]
Travis failure in Kafka test. Does not look related to this change and possibly timing related. Would trigger a rebuild

```
org.scalatest.exceptions.TestFailedException: 1 was not equal to 2Open stacktrace
Send message to topic.
Successfully sent message to topic: KafkaConnectorTestTopic-0@5
Receiving message from topic.
Received 1. Took 15 msec: List(Wed Feb 28 10:59:19 UTC 2018)
Send message to topic.
Successfully sent message to topic: KafkaConnectorTestTopic-0@6
Receiving message from topic.
Received 1. Took 16 msec: List(Wed Feb 28 10:59:19 UTC 2018)
Send message to topic.
Successfully sent message to topic: KafkaConnectorTestTopic-0@7
Receiving message from topic.
Received 1. Took 30 msec: List(Wed Feb 28 10:59:19 UTC 2018)
Send message to topic.
Successfully sent message to topic: KafkaConnectorTestTopic-0@8
Receiving message from topic.
Received 1. Took 5 msec: List(Wed Feb 28 10:59:30 UTC 2018)
```

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala (hunk 1 above)
Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: May be we deal with this separately as I see this distinction between CouchDB and Cloudant in few places like

1. CloudantRestClient
2. Vagrantfile
3. CouchDbRestClientTests

**Comment by chetanmeh** [comment]
Updated the PR to now remove usage of WhiskConfig from the ArtifactStoreProvider api completely. Now database names are also captured via the application.conf and config keys are derived on the basis of entity class name

@markusthoemmes  Let me know if any other aspect needs to be addressed as part of this PR

File: tests/src/test/scala/whisk/core/controller/test/ActivationsApiTests.scala [hunk 2]
```diff
@@ -526,18 +526,33 @@ class ActivationsApiTests extends ControllerTestCommon with WhiskActivationsApi
...
         WhiskDocumentReader,
         system,
         logging,
         materializer)
     implicit val tid = transid()
-    val entity = BadEntity(namespace, EntityName(ActivationId().toString))
-    put(activationStore, entity)
 
-    Get(s"$collectionPath/${entity.name}") ~> Route.seal(routes(creds)) ~> check {
+    class BadActivation(override val namespace: EntityPath,
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/controller/test/ActivationsApiTests.scala]: Why are these changes needed in this context?

File: tests/src/test/scala/ha/ShootComponentsTests.scala [hunk 3]
```diff
@@ -67,13 +70,13 @@ class ShootComponentsTests
...
   val dbHostsList = WhiskProperties.getDBHosts
-  val dbPort = WhiskProperties.getProperty(WhiskConfig.dbPort)
-  val dbUsername = WhiskProperties.getProperty(WhiskConfig.dbUsername)
-  val dbPassword = WhiskProperties.getProperty(WhiskConfig.dbPassword)
-  val dbPrefix = WhiskProperties.getProperty(WhiskConfig.dbPrefix)
-  val dbWhiskAuth = WhiskProperties.getProperty(WhiskConfig.dbAuths)
+  val dbPort = dbConfig.port
+  val dbUsername = dbConfig.username
+  val dbPassword = dbConfig.password
+  val dbWhiskAuth = dbConfig.databases.get("WhiskAuth").get
```
Comment (markusthoemmes) [inline tests/src/test/scala/ha/ShootComponentsTests.scala]: Can this be `dbConfig.databaseFor[WhiskAuth]`?

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 4]
```diff
@@ -21,37 +21,49 @@ import akka.actor.ActorSystem
...
+case class CouchDbConfig(provider: String,
+                         protocol: String,
+                         host: String,
+                         port: Int,
+                         username: String,
+                         password: String,
+                         databases: Map[String, String]) {
+  assume(Set(protocol, host, username, password).forall(_.nonEmpty), "At least one expected property is missing")
+
+  def getDatabaseName(entityClass: Class[_]): String = {
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: This is a style type comment:

I'd propose a different name/API for this method. In the vein of the `ClassTag` used in `makeStore`, we can use `ClassTag` here as well to make a neat call like `databaseFor[WhiskAuth]`. Proposed diff attached. WDYT?

```diff
diff --git a/common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala b/common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala
index 7e01c828..c10436b5 100644
--- a/common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala
+++ b/common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala
@@ -36,10 +36,11 @@ case class CouchDbConfig(provider: String,
                          databases: Map[String, String]) {
   assume(Set(protocol, host, username, password).forall(_.nonEmpty), "At least one expected property is missing")
 
-  def getDatabaseName(entityClass: Class[_]): String = {
-    databases.get(entityClass.getSimpleName) match {
+  def databaseFor[D](implicit tag: ClassTag[D]): String = {
+    val entityType = tag.runtimeClass.getSimpleName
+    databases.get(entityType) match {
       case Some(name) => name
-      case None       => throw new IllegalArgumentException(s"Database name mapping not found for $entityClass")
+      case None       => throw new IllegalArgumentException(s"Database name mapping not found for $entityType")
     }
   }
 }
@@ -63,7 +64,7 @@ object CouchDbStoreProvider extends ArtifactStoreProvider {
       dbConfig.port,
       dbConfig.username,
       dbConfig.password,
-      dbConfig.getDatabaseName(implicitly[ClassTag[D]].runtimeClass),
+      dbConfig.databaseFor[D],
       useBatching)
   }
 }

```

File: common/scala/src/main/resources/application.conf [hunk 5]
```diff
@@ -106,6 +106,23 @@ whisk {
         activations-ddoc = "whisks.v2.1.0"
         activations-filter-ddoc = "whisks-filters.v2.1.0"
     }
+
+    # CouchDB related configuration
```
Comment (markusthoemmes) [inline common/scala/src/main/resources/application.conf]: How do these settings relate to the design-document settings above? Those are couchdb specific as well, aren't they?

File: tests/src/test/scala/ha/ShootComponentsTests.scala (hunk 3 above)
Comment (chetanmeh) [reply tests/src/test/scala/ha/ShootComponentsTests.scala]: Problem here is `WhiskAuth` is scoped to `core` package and test lives outside of that package. So need to use the string value explicitly

File: common/scala/src/main/resources/application.conf (hunk 5 above)
Comment (chetanmeh) [reply common/scala/src/main/resources/application.conf]: The design docs are used by layers above `ArtifactStore` and hence not internal to it. Infact the view names are used for other [custom stores][1] to determine how to frame the queries.

So they need to be kept separately 

[1]: https://github.com/chetanmeh/openwhisk-mongo/blob/master/common/mongo/src/main/scala/whisk/core/database/mongo/MongoViewMapper.scala

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala (hunk 4 above)
Comment (chetanmeh) [reply common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: Good point. Updated PR with above patch

File: tests/src/test/scala/whisk/core/controller/test/ActivationsApiTests.scala (hunk 2 above)
Comment (chetanmeh) [reply tests/src/test/scala/whisk/core/controller/test/ActivationsApiTests.scala]: The test here was validating that deserialization issue triggers 500 response. For this it injected a WhiskEntity in Activation table by "misusing" the dbName parameter

```
 val activationStore = SpiLoader
      .get[ArtifactStoreProvider]
      .makeStore[WhiskEntity](whiskConfig, _.dbActivations)(
        WhiskEntityJsonFormat,
```
Now as the db name is now mapped implicitly by entity type class this would not work. So test has been modified to use a `BadActivation` which emits wrong json and thus triggers the deserialization fault

Comment (markusthoemmes) [reply tests/src/test/scala/whisk/core/controller/test/ActivationsApiTests.scala]: Ah got it, could you add a comment to the `toJson` method stating that it breaks the deserialization by removing the subject entry?

**Comment by chetanmeh** [comment]
Last build failed due to low memory. Triggered a rebuild

```
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 139984896 bytes for committing reserved memory.
# An error report file with more information is saved as:
Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00000000d0b80000, 139984896, 0) failed; error='Cannot allocate memory' (errno=12)
# /home/travis/build/apache/incubator-openwhisk/tests/hs_err_pid19764.log
```

**Comment by markusthoemmes** [comment]
PG1 2677 running.

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala (hunk 1 above)
Comment (rabbah) [reply common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: This is vestigial at this point and can be removed in separate PR.

**Comment by markusthoemmes** [comment]
@chetanmeh I'll fix the conflicts.

**Comment by chetanmeh** [comment]
Thanks @markusthoemmes . Let me know if it needs any further change
